cmd/duffle/relocate_test.go

```diff
@@ -0,0 +1,183 @@
+	relocatedImageDigestB = "sha256:deadbeef892aeccb7e142557fe746ccd0a8f736a747c195ef04c9f3f0f0bbd49"
+)
+
+func TestRelocateFileToFilePreservingDigests(t *testing.T) {
+	relocateFileToFile(t, true, nil)
+}
+
+func TestRelocateFileToFileMutatingDigests(t *testing.T) {
+	relocateFileToFile(t, false, fmt.Errorf("digest of image %s not preserved: old digest %s; new digest %s", originalImageNameB, originalImageDigestB, relocatedImageDigestB))
+}
+
+func relocateFileToFile(t *testing.T, preserveDigest bool, expectedErr error) {
+
+	duffleHome, err := ioutil.TempDir("", "dufflehome")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(duffleHome)
+
+	work, err := ioutil.TempDir("", "relocatetest")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(work)
+
+	outputBundle := filepath.Join(work, "relocated.json")
+
+	cmd := &relocateCmd{
+		inputBundle:  "testdata/relocate/bundle.json",
+		outputBundle: outputBundle,
+
+		repoPrefix:         testRepositoryPrefix,
+		inputBundleIsFile:  true,
+		outputBundleIsFile: true,
+
+		home: home.Home(duffleHome),
+		out:  ioutil.Discard,
+
+		mapping: func(repoPrefix string, originalImage image.Name) image.Name {
+			if repoPrefix != testRepositoryPrefix {
+				t.Fatalf("Unexpected repository prefix %s", repoPrefix)
+			}
+			return testMapping(originalImage, t)
+		},
+		registryClient: &mockRegClient{
+			copyStub: func(source image.Name, target image.Name) (image.Digest, error) {
+				oinA, err := image.NewName(originalImageNameA)
+				if err != nil {
+					t.Fatal(err)
+				}
+				oinB, err := image.NewName(originalImageNameB)
+				if err != nil {
+					t.Fatal(err)
+				}
+				switch source {
+				case oinA:
+					if target.String() == relocatedImageNameA {
+						return image.NewDigest(imageDigestA)
+					}
+				case oinB:
+					if target.String() == relocatedImageNameB {
+						if preserveDigest {
+							return image.NewDigest(originalImageDigestB)
+						}
+						// check behaviour if digest is modified, even though this is not normally expected
+						return image.NewDigest(relocatedImageDigestB)
+					}
+				default:
+					t.Fatalf("unexpected source %v", source)
+				}
+				t.Fatalf("unexpected mapping from %v to %v", source, target)
+				return image.EmptyDigest, nil // unreachable
+			},
+		},
+	}
+
+	if err := cmd.run(); (err != nil && expectedErr != nil && err.Error() != expectedErr.Error()) ||
+		((err == nil || expectedErr == nil) && err != expectedErr) {
+		t.Fatalf("unexpected error %v (expected %v)", err, expectedErr)
+	}
+
+	if expectedErr != nil {
+		return
+	}
+
+	// check output bundle
+	bundleFile, err := resolveBundleFilePath(outputBundle, "", true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bun, err := loadBundle(bundleFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = bun.Validate(); err != nil {
+		t.Fatal(err)
+	}
+
+	assertImage := func(t *testing.T, i string, expectedOriginalImageName string, expectedImageName string) {
```

Comment:
If I read this correctly, the only captured element here is the bundle `bun`. I guess this functor can be extracted as-is as a helper function, by just adding the bundle as a parameter.

Reply:
Yes, that's right, only `bun` is captured. However, for consistency, `t` should be captured too. The reason for using a function (not actually a functor, note) here is that it reduces the repetition of arguments at the call sites. I guess it's a matter of personal preference.